extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java

```diff
@@ -0,0 +1,232 @@
+     * Parse out a URI string and get a GuacamoleConfiguration
+     * from that string, or an exception if the parsing fails.
+     *
+     * @param uri
+     *     The string form of the URI to be parsed.
+     *
+     * @return
+     *     A GuacamoleConfiguration using a combination of the parsed
+     *     URI values and default values when not specified in the
+     *     URI.
+     *
+     * @throws GuacamoleException
+     *     When an error occurs parsing the URI.
+     */
+    public static GuacamoleConfiguration getConfiguration(String uri)
+            throws GuacamoleException {
+
+        // Parse the URI object from provided string.
+        URI qcUri;
+        try {
+            qcUri = new URI(uri);
+        }
+        catch (URISyntaxException e) {
+            throw new GuacamoleClientException("Invalid URI Syntax", e);
+        }
+
+        // Break out individual components of the URI.
+        String protocol = qcUri.getScheme();
+        String host = qcUri.getHost();
+        int port = qcUri.getPort();
+        String userInfo = qcUri.getUserInfo();
+        String query = qcUri.getQuery();
+
+        // Generate a new GuacamoleConfiguration
+        GuacamoleConfiguration qcConfig = new GuacamoleConfiguration();
+
+        // Check for provided protocol or use default
+        if (protocol != null && !protocol.isEmpty())
+            qcConfig.setProtocol(protocol);
+        else
+            qcConfig.setProtocol(DEFAULT_URI_PROTOCOL);
+
+        // Check for provided port number
+        if (port > 0)
+            qcConfig.setParameter("port", Integer.toString(port));
+
+        // Check for provided host or use default
+        if (host != null && !host.isEmpty())
+            qcConfig.setParameter("hostname", host);
+        else
+            qcConfig.setParameter("hostname", DEFAULT_URI_HOST);
+
+        // Look for extra query parameters and parse them out.
+        if (query != null && !query.isEmpty()) {
+            try {
+                Map<String, String> queryParams = parseQueryString(query);
+                if (queryParams != null)
+                    for (Map.Entry<String, String> entry: queryParams.entrySet())
+                        qcConfig.setParameter(entry.getKey(), entry.getValue());
+            }
+            catch (UnsupportedEncodingException e) {
+                throw new GuacamoleServerException("Unexpected lack of UTF-8 encoding support.", e);
+            }
+        }
+
+        // Look for the username and password and parse them out.
+        if (userInfo != null && !userInfo.isEmpty()) {
+
+            Matcher userinfoMatcher = userinfoPattern.matcher(userInfo);
+            if (userinfoMatcher.matches()) {
+                String username = userinfoMatcher.group(USERNAME_GROUP);
+                String password = userinfoMatcher.group(PASSWORD_GROUP);
+
+                if (username != null && !username.isEmpty())
+                    qcConfig.setParameter("username", username);
```

Comment:
I see `parseQueryString()` automatically handling URL decoding using `URLDecoder.decode()`. Does that need to happen here, too, with the username and password?

Reply:
I've added the `URLDecoder.decode()` method to this, and I also broke out the parsing of the username/password into its own method, as it was gaining the same level of complexity that the query string parser had.